torch/lib/THCUNN/SpatialDepthwiseConvolution.cu

```diff
@@ -54,18 +54,18 @@ __global__ void spatialDepthwiseConvolutionUpdateOutput(
         if ((h_in >= 0) && (h_in < inputHeight) && (w_in >= 0) && (w_in < inputWidth)) {
           const IndexType offset = ((n * inputChannels + inputChannel) * inputHeight + h_in) *
                                     inputWidth + w_in;
-          value = THCNumerics<T>::add(
+          value = THCNumerics<AccT>::add(
             value,
-            THCNumerics<T>::mul(weight.data()[weightOffset], input.data()[offset]));
+            ScalarConvert<T, AccT>::to(THCNumerics<T>::mul(weight.data()[weightOffset], input.data()[offset])));
```

Comment:
It will be better to ScalarConvert weight.data()[weightOffset] and input.data()[offset] to AccT (no op for float and double, will be done anyway by THCNumerics<T>::mul for halfs anyway), and just multiply them (or do THCNumerics<AccT>::mul). The way it is now, for halfs there's an unnecessary conversion to half (in THCNumerics<T>::mul) and back (in ScalarConvert<T,AccT>), that is also losing accuracy.

Reply:
Sure, I did this for all of the kernels, but HalfPrecision still has pretty high errors (i.e. it causes test failures), is this expected @ngimel ?

Reply:
What errors are you getting? In cudnn.torch tolerance is set fairly high for half https://github.com/soumith/cudnn.torch/blob/R7/test/test.lua#L13-L14, I did not think it was necessary, but may be it is...

Reply:
A little less than 0.01. I will update the error bounds for half.

Reply:
1e-2 sounds like a reasonable error bound for half (I think it is set for some pytorch tests). If there are some flaky tests, it could be increased, but probably not to the torch levels, they are too high.